pkg/credentialprovider/plugin/plugin.go

```diff
@@ -0,0 +1,393 @@
+	cache cache.Store
+	// defaultCacheDuration is the default duration credentials are cached in-memory if the auth plugin
+	// response did not provide a cache duration for credentials.
+	defaultCacheDuration time.Duration
+
+	// plugin is the exec implementation of the credential providing plugin.
+	plugin Plugin
+}
+
+// cacheEntry is the cache object that will be stored in cache.Store.
+type cacheEntry struct {
+	key         string
+	credentials credentialprovider.DockerConfig
+	expiresAt   time.Time
+}
+
+// cacheKeyFunc extracts AuthEntry.MatchKey as the cache key function for the plugin provider.
+func cacheKeyFunc(obj interface{}) (string, error) {
+	key := obj.(*cacheEntry).key
+	return key, nil
+}
+
+// cacheExpirationPolicy defines implements cache.ExpirationPolicy, determining expiration based on the expiresAt timestamp.
+type cacheExpirationPolicy struct{}
+
+// IsExpired returns true if the current time is after cacheEntry.expiresAt, which is determined by the
+// cache duration returned from the CRI auth plugin response.
+func (c *cacheExpirationPolicy) IsExpired(entry *cache.TimestampedEntry) bool {
+	return time.Now().After(entry.Obj.(*cacheEntry).expiresAt)
+}
+
+// Provide returns a credentialprovider.DockerConfig based on the credentials returned
+// from cache or the exec plugin.
+func (p *pluginProvider) Provide(image string) credentialprovider.DockerConfig {
+	if !p.isImageAllowed(image) {
+		return credentialprovider.DockerConfig{}
+	}
+
+	p.Lock()
+	defer p.Unlock()
+
+	cachedConfig, found, err := p.getCachedCredentials(image)
+	if err != nil {
+		klog.Errorf("Failed to get cached docker config: %v", err)
+		return credentialprovider.DockerConfig{}
+	}
+
+	if found {
+		return cachedConfig
+	}
+
+	// fetch credentials from the exec plugin if the image is allowed and not already matched against cached objects
+	response, err := p.plugin.ExecPlugin(image)
+	if err != nil {
+		klog.Errorf("Failed getting credential from external registry credential provider: %v", err)
+		return credentialprovider.DockerConfig{}
+	}
+
+	if len(response.Auth) == 0 {
+		return credentialprovider.DockerConfig{}
+	}
+
+	var cacheKey string
+	switch cacheKeyType := response.CacheKeyType; cacheKeyType {
+	case criauth.ImagePluginCacheKeyType:
+		cacheKey = image
+	case criauth.RegistryPluginCacheKeyType:
+		registry, err := parseRegistry(image)
+		if err != nil {
+			klog.Errorf("error parsing registry from image %s: %v", image, err)
+			return credentialprovider.DockerConfig{}
+		}
+
+		cacheKey = registry
+	case criauth.GlobalPluginCacheKeyType:
+		cacheKey = globalCacheKey
+	}
+
+	dockerConfig := make(credentialprovider.DockerConfig, 0)
+	for matchImage, authConfig := range response.Auth {
+		dockerConfig[matchImage] = credentialprovider.DockerConfigEntry{
+			Username: authConfig.Username,
+			Password: authConfig.Password,
+			Email:    "plugin@criauth.k8s.io", // email is deprecated, this is just a dummy email
```

Comment:
hmm... would prefer this not be a real domain name... do we have to put a value here?

Reply:
> do we have to put a value here?

Every plugin puts a dummy value here which is why I added it. If we can get away with removing it that would be great

Reply:
(I'm going to remove it and see if I can break any providers)

Reply:
something with an IANA dummy domain (like example.com) would be preferable